### PR TITLE
ddns-scripts: fix regression added by version 2.8.0

### PR DESCRIPTION
**Pull request by feckert**
Maintainer: me / @Ansuel 
Compile tested: not needed only script changes
Run tested: x86_64, APU3, openwrt master

Description:

- Rename package for services installation https://github.com/openwrt/packages/pull/13402#issuecomment-699025971
- Rename packages for special services installation https://github.com/openwrt/packages/pull/13402#issuecomment-699025971
- Remove installation list file for custom providers this file gets installed by the new ddns script
- Refactoring new ddns script to save additional providers into custom directory
- Fix installation json files for special providers

The following issues should get resolved with this P/R.
https://github.com/openwrt/packages/issues/13495
https://lists.openwrt.org/pipermail/openwrt-devel/2020-September/031507.html

@Ansuel I have not add the dependency ddns-scripts-services to ddns-scripts for now. When I do that the menuconfig does not look clean! Could you verify this.

### CODE REVIEW

**Comment by Ansuel** [comment]
What problem do you have with the deps?
```
define Package/ddns-scripts
  $(call Package/ddns-scripts/Default)
  TITLE:=Dynamic DNS Client scripts (with IPv6 support)
  DEPENDS:=+ddns-scripts-services
endef

define Package/ddns-scripts/description
  Dynamic DNS Client scripts (with IPv6 support)
  A highly configurable set of scripts for doing dynamic dns updates.
  - IPv6 support
  - DNS server support
  - Glue Record support (require BIND host or KNOT host)
  - DNS requests via TCP
  - Proxy server support
  - log file support
  - support to run once
  Version: $(PKG_VERSION)-$(PKG_RELEASE)
  Info   : https://openwrt.org/docs/guide-user/services/ddns/client
endef

define Package/ddns-scripts/conffiles
/etc/config/ddns
endef


define Package/ddns-scripts-services
  $(call Package/ddns-scripts/Default)
  TITLE:=Common ddns providers
#   DEPENDS:=ddns-scripts
endef
```
Immagine 2020-09-15 000925

File: net/ddns-scripts/Makefile [hunk 1]
```diff
@@ -224,14 +222,24 @@ exit 0
...
+	rm $(1)/usr/share/ddns/services/digitalocean.com-v2.json
+	rm $(1)/usr/share/ddns/services/no-ip.com.json
+	rm $(1)/usr/share/ddns/services/bind-nsupdate.json
+	rm $(1)/usr/share/ddns/services/route53-v1.json
+	rm $(1)/usr/share/ddns/services/cnkuai.cn.json
 endef
 
 
-define Package/ddns-scripts_cloudflare.com-v4/install
+define Package/ddns-scripts-cloudflare/install
```
Comment (Ansuel) [inline net/ddns-scripts/Makefile:242]: I would keep the same name and just change the '_' to '-'. People would get confuse and AFAIK there could be various version of the same script.

**Comment by Ansuel** [comment]
I was thinking of renaming the 'services' dir to 'default'. This would make it more clear that in that dir there are json installed by the opkg and should not be touched...
Aside from this. Can we keep the list file on the repo but not install it?

File: net/ddns-scripts/Makefile (hunk 1 above)
Comment (feckert) [reply net/ddns-scripts/Makefile:242]: > I would keep the same name and just change the '_' to '-'. People would get confuse and AFAIK there could be various version of the same script.

I do not understand that. We can add more scripts in the future. So I think it's legitimate if we keep the name more generic. The json file is still that way. So the uci parameter will not change.

**Comment by feckert** [comment]
> I was thinking of renaming the 'services' dir to 'default'. This would make it more clear that in that dir there are json installed by the opkg and should not be touched...

Sounds good :+1:. I will change that.

> Aside from this. Can we keep the list file on the repo but not install it?

This should fit. I have only removed the installation call and the file is still in the repo

> I have not add the dependency ddns-scripts-services to ddns-scripts for now. When I do that the menuconfig does not look clean! Could you verify this.

forget it. That has been cleared up I have rebuilt and then it worked. But thanks

**Comment by Ansuel** [comment]
> This should fit. I have only removed the installation call and the file is still in the repo

Didn't notice... Will update the luci app pr with the default dir so we can merge both... What do you think?

**Comment by feckert** [comment]
@Ansuel I have to update this P/R with another commit to read also the custom json values.
For now the [read function](https://github.com/openwrt/packages/blob/master/net/ddns-scripts/files/usr/lib/ddns/dynamic_dns_functions.sh#L332) does only read the files from the services directory (In the future this directory is changed in this P/R to default, as you suggested).

But we could not read the custom directory if we have installed a provider in this directory for now.
Any solution to fix this?

My suggestions are:

- Read also in another call the values from the custom directory if we do not find the json file in the default directory
- Check first the custom directory and then the default directory

An if we find no json then we gave up?

I would use the second suggest. So we have always use the newest json information if we try to read first the custom directory.

**Comment by Ansuel** [comment]
> I would use the second suggest.

Yes the default json should be the fallback if custom is not provided. Maybe add a comment somewhere to make this clear (that any json have priority than default?)

File: net/ddns-scripts/Makefile (hunk 1 above)
Comment (Ansuel) [reply net/ddns-scripts/Makefile:242]: The old name directly told what service the pkg installs (without reading the description). But now that I think about it's better to have a more generic name (if more version of the same service will comes up, the generic special pkg will install all of them). Sorry for the mess.

**Comment by snakwu** [comment]
@feckert @Ansuel   
什么时候可以合并？

**Comment by neheb** [comment]
```
 In net/ddns-scripts/files/usr/lib/ddns/dynamic_dns_updater.sh line 48:
	printf %s\n "$MYPROG: $@" >&2
                               ^-- SC2145: Argument mixes string and array. Use * or separate argument.
```

**Comment by feckert** [comment]
> ```
>  In net/ddns-scripts/files/usr/lib/ddns/dynamic_dns_updater.sh line 48:
> 	printf %s\n "$MYPROG: $@" >&2
>                                ^-- SC2145: Argument mixes string and array. Use * or separate argument.
> ```

@neheb Thanks for your advice. But I don't want to deal with the shellcheck things in this PR. Now I have to make sure that the JSONs work properly. I want to update some other things in the package. The restructuring of the package is not yet complete. There is still a lot to be done.

**Comment by feckert** [comment]
@Ansuel I would merge the changes now. If it works for you, too.

**Comment by Ansuel** [comment]
Sorry for the delay... I tested now and it seems to work all correctly

**Comment by ivanich** [comment]
Not sure how but this bricks my edgerouterx, had to recovery via tftp. When I remove this commit from packages repo it works fine

**Comment by feckert** [comment]
@ivanich That are only script changes. So what did you exactly do. This changes were recently committed so i do not think this has already picked up by the buildbot and that this package are available in the snapshot for installation.

**Comment by ivanich** [comment]
@feckert I make my own builds for 3 routers including edgerouter for a quite long time. As I said before build with this commit reverted works fine. I even tried to disable all ddns staff in the config but it still wont boot.

**Comment by ivanich** [comment]
@feckert I simply build openwrt from sources as I I always do. Basically build I did before your commit works fine, with your commit router won't boot i.e. no ethernet links on all lan ports except WAN.

**Pull request merged**